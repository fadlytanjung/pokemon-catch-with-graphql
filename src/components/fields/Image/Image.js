import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.scoped.css';

export default function Image(props) {
  const { className, input, src } = props;
  const [ preview, setPreview ] = useState('');
  const classes = [
    styles.root,
    className,
  ].filter(Boolean).join(' ');

  const onChange = e => {
    const { files } = e.target;
    const reader = new FileReader();
    reader.onload = e => {
      setPreview(e.target.result);
      input.onChange(files[0]);
    };
    reader.readAsDataURL(files[0]);
  };

  return (
    <label className={classes} style={{ backgroundImage: `url('${preview ? preview : src}')` }}>
      <input accept="image/png, image/jpeg, image/jpg" onChange={onChange} style={{ display: 'none' }} type="file" />
      <input {...input} type="hidden" />
    </label>
  );
}

Image.defaultProps = {
  className: '',
  input: {},
  src: '',
};

Image.propTypes = {
  className: PropTypes.string,
  input: PropTypes.object,
  src: PropTypes.string,
};
