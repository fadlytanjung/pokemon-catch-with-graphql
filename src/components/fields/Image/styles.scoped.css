.root {
  background-position: contain;
  background-repeat: no-repeat;
  background-size: cover;
  cursor: pointer;
  display: inline-block;
}
