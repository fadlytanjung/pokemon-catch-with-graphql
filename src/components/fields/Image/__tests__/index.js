import React, { useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Image from '../Image';

describe('src/components/fields/Image', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Image />);
    expect(tree).toMatchSnapshot();
  });

  test('render with preview & onChange', () => {
    const setPreview = jest.fn();
    useState.mockImplementationOnce(() => ['preview', setPreview]);

    delete global.FileReader;
    const readAsDataURL = jest.fn();
    const reader = { readAsDataURL };
    global.FileReader = () => reader;

    const { defaultProps } = Image;
    const input = { onChange: jest.fn() };
    const result = Image({ ...defaultProps, input });
    expect(result.props.style.backgroundImage).toBe('url(\'preview\')');

    const target = { files: ['file'], result: 'result' };
    result.props.children[0].props.onChange({ target });
    reader.onload({ target });
    expect(setPreview).toHaveBeenCalledWith('result');
    expect(readAsDataURL).toHaveBeenCalledWith('file');
  });
});
