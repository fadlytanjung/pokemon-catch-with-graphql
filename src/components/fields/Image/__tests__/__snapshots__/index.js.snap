// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/fields/Image render 1`] = `
<label
  className=""
  style={
    Object {
      "backgroundImage": "url('')",
    }
  }
>
  <input
    accept="image/png, image/jpeg, image/jpg"
    onChange={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
    type="file"
  />
  <input
    type="hidden"
  />
</label>
`;
